dev/OAuth/AuthFailure.cpp

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+#include <pch.h>
+#include "common.h"
+
+#include "AuthFailure.h"
+
+#include <Microsoft.Security.Authentication.OAuth.AuthFailure.g.cpp>
+
+using namespace std::literals;
+using namespace winrt::Microsoft::Security::Authentication::OAuth;
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Foundation::Collections;
+
+namespace winrt::Microsoft::Security::Authentication::OAuth::implementation
+{
+    AuthFailure::AuthFailure(const Uri& responseUri)
+    {
+        std::map<winrt::hstring, winrt::hstring> additionalParams;
+
+        auto parseComponents = [&](const winrt::hstring& str) {
+            if (str.empty())
+            {
+                return; // Avoid unnecessary construction/activation
+            }
+
+            for (auto&& entry : WwwFormUrlDecoder(str))
+            {
+                auto name = entry.Name();
+                if (name == L"error"sv)
+                {
+                    m_error = entry.Value();
+                }
+                else if (name == L"error_description"sv)
+                {
+                    m_errorDescription = entry.Value();
+                }
+                else if (name == L"error_uri"sv)
+                {
+                    m_errorUri = Uri(entry.Value());
+                }
+                else if (name == L"state"sv)
+                {
+                    m_state = entry.Value();
+                }
+                else
+                {
+                    additionalParams.emplace(std::move(name), entry.Value());
+                }
+            }
+        };
+
+        parseComponents(responseUri.Query());
+        parseComponents(fragment_component(responseUri));
```

Comment:
Why you need to call the parseComponents two times?

Reply:
If we have a URI like https://example.com/callback#access_token=abc123, calling fragment_component on this URI would return access_token=abc123.
https://www.rfc-editor.org/rfc/rfc6749#section-4.2.2
![image](https://github.com/user-attachments/assets/42a54abd-3825-4b7a-a50a-9c4372ce8a3b)